### ISSUE TITLE
TSecAuthService#withAuthorization and HierarchyAuth

### ISSUE THREAD

**Issue by fmonniot** [jmcardon/tsec#160]
Hello there,

First of all let me thank you for this incredible library, long live to it !

Back to the issue at hand, I was playing with authorization in http4s and though that `tsec.authorization.HierarchyAuth` was a good scheme. So fearlessly, I changed the BasicRBAC from [the examples](https://github.com/jmcardon/tsec/blob/master/examples/src/main/scala/http4sExamples/ExampleAuthHelpers.scala). Which didn't worked out, because `HierarchyAuth.apply` return an `F[HierarchyAuth[_, _, _, _]]` and `tsec.authenticationTSecAuthService.withAuthorization` take the `Authorization` directly.

So here is why I opened this issue: do you think it would make sense to offer a method which take a `F[Authorization[_, _, _]]` directly ? Something like:
```scala
def withAuthorizationF[I, A, F[_]](auth: F[Authorization[F, I, A]])
                                  (pf: PartialFunction[SecuredRequest[F, I, A], F[Response[F]]])
                                  (implicit F: Monad[F]): TSecAuthService[I, A, F]
```

In my own project I ended up implemented a similar method (see below), and I'm wondering if that would have its place in the core project ?
```scala
// Use a parameterized TAuth because F is invariant 
def withAuthorizationF[I, A, F[_], TAuth <: Authorization[F, I, A]](fAuth: F[TAuth])(
  pf: PartialFunction[SecuredRequest[F, I, A], F[Response[F]]]
)(implicit F: Monad[F]): TSecAuthService[I, A, F] =
  Kleisli { req: SecuredRequest[F, I, A] =>
    OptionT.liftF(fAuth)
      .flatMap { auth =>
        auth
          .isAuthorized(req)
          .flatMap(
            _ => pf.andThen(OptionT.liftF(_)).applyOrElse(req, Function.const(OptionT.none))
          )
      }
  }
```

Thanks

**Comment by jmcardon** [comment]
Hi. I'm glad you like the library!

PRs are always welcome, sure. :+1:

**Comment by jmcardon** [comment]
If the purpose of this is to supply authorization and extraction of it separately from the user object I'm all for it. If that's something people want, we're still pre-RC1 so feel free.

**Comment by fmonniot** [comment]
Ok, I'll submit a PR then.
Not sure what you meant with your second comment though.

**Comment by fmonniot** [comment]
Closed via #161
